ui/src/Main.purs

```diff
@@ -77,13 +80,18 @@ update s = case _ of
         messageHandler = \msgStr -> do
           let
             jsonBlob = parseJson msgStr
-            devices = Devices.decodeDevices =<< jsonBlob
+            devices = fromRight (Devices.mkFailedParse jsonBlob) $
+                        Devices.decodeDevices <$> jsonBlob
             deviceState = DeviceState.decodeDeviceState =<< jsonBlob
-            groups = Groups.decodeGroups =<< jsonBlob
+            groups = case devices of
+              Decoded DecodingSucceeded devices' ->
+                Groups.decodeGroups devices'.devices =<< jsonBlob
+              _ ->
+                Groups.decodeGroups s.devices.devices =<< jsonBlob
```

Comment:
I guess we should be doing this any time we get a jsonBlob, since we don't know if it's for Groups or not, and we should rebuild groups with existing Devices if we get a Groups update.